### PR TITLE
Fixed behavior of insert track queue in `AnimationTrackEditor` is unstable

### LINKED ISSUE #51415: Inserting multiple tracks in `AnimationTrackEditor` is broken
### Godot version

4.0.dev

### System information

MacOS 11.4

### Issue description

`_query_insert` in AnimationTrackEditor used to be able to insert multiple tracks, but now it doesn't work. From my investigation, it determines if the insertions were done on the same frame and summarizes the operations. It seems that the renderer has changed and is no longer working. Regression, but I think it is not good implementation of "determine if it's the same frame" in the first place, so it needs to be improved.

### Steps to reproduce

Try to insert multiple animations at the same time by iteration.

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by TokageItLab**
Fixed #51415.

The queue for insert track in `AnimationTrackEditor` gets the drawing frame of the Engine and determines if the operation was done on the same frame. Why do we implement it this way? This implementation seems to be special compared to other queues.

There does not seem to be any code anywhere that uses the insert track queue, but when we actually try to use it, it does not work well in 4.0 (at least it was working well in 3.3) where the renderer has changed significantly.

I made it so that the queue creation and commit is done manually.

### CODE REVIEW

**Comment by TokageItLab** [comment]
I found a problem with undo and am working on fixing it.

**Comment by TokageItLab** [comment]
I assume that this queue was only for BezierTrack, and if we try to use it from outside of AnimationTrackEditor, it works fine in 3.3 and earlier, but not in 4.0. Anyway, I think this is not a good way to implement queue, and it is actually causing problems in the implementation of #45699, so I fixed it.

File: editor/animation_track_editor.cpp [hunk 1]
```diff
@@ -3383,74 +3369,82 @@ void AnimationTrackEditor::_query_insert(const InsertData &p_id) {
...
-			if (insert_data[i].type != Animation::TYPE_VALUE && insert_data[i].type != Animation::TYPE_BEZIER) {
-				all_bezier = false;
-			}
-
-			if (insert_data[i].track_idx == -1) {
-				++num_tracks;
-			}
+	// organize insert data
+	int num_tracks = 0;
+	String last_track_query = "";
```
Comment (akien-mga) [inline editor/animation_track_editor.cpp:3374]: `String` is an empty string by default, no need to initialize.

File: editor/animation_track_editor.cpp [hunk 2]
```diff
@@ -3383,74 +3369,82 @@ void AnimationTrackEditor::_query_insert(const InsertData &p_id) {
...
+	for (const InsertData &E : insert_data) {
+		//prevent insertion of multiple tracks
+		if (E.path == p_id.path) {
+			return; //already inserted a track for this on this frame
+		}
+	}
+
+	insert_data.push_back(p_id);
+
+	// Without queue, commit immidiately
```
Comment (akien-mga) [inline editor/animation_track_editor.cpp:3441]:
```suggestion
	// Without queue, commit immediately.
```

File: editor/animation_track_editor.cpp [hunk 3]
```diff
@@ -3383,74 +3369,82 @@ void AnimationTrackEditor::_query_insert(const InsertData &p_id) {
...
-		return;
+void AnimationTrackEditor::_query_insert(const InsertData &p_id) {
+	if (!insert_queue) {
+		insert_data.clear();
 	}
 
+	for (const InsertData &E : insert_data) {
+		//prevent insertion of multiple tracks
+		if (E.path == p_id.path) {
+			return; //already inserted a track for this on this frame
```
Comment (akien-mga) [inline editor/animation_track_editor.cpp:3435]: While at it, could improve the comment style:
```suggestion
		// Prevent insertion of multiple tracks.
		if (E.path == p_id.path) {
			return; // Already inserted a track for this on this frame.
```

**Comment by TokageItLab** [comment]
I fixed the part that was reviewed.

**Comment by TokageItLab** [comment]
@JFonS As I mentioned in #45699, if this implementation is not done, inserting bone animation will not work properly and the implementation of SkeletonGizmo is halted, so if you can, it would be helpful that you review and approve this PR.

**Comment by JFonS** [comment]
I don't know much about the animation track system, so I can't properly review this PR, sorry.

**Comment by TokageItLab** [comment]
@fire Here is video demo.

This is the behavior in the current master.

https://user-images.githubusercontent.com/61938263/130363462-313b11f5-7625-401d-8b82-4fc17da5c34f.mov

If I do `track_insert_key()` from within `SkeletonEditorPlugin` with the number of bones, only one track is inserted. Also, you can see that the timing of the popup warning is wrong.

Thus, for example, I provided one frame of grace for the judging.

```
void AnimationTrackEditor::_query_insert(const InsertData &p_id) {
	if (insert_frame != Engine::get_singleton()->get_frames_drawn()) {
		//clear insert list for the frame if frame changed
		if (insert_confirm->is_visible()) {
			return; //do nothing
		}
		insert_data.clear();
		insert_query = false;
	}
	insert_frame = Engine::get_singleton()->get_frames_drawn() + 1; // <- one frame of grace
```

https://user-images.githubusercontent.com/61938263/130363473-293ac9f2-79e0-43eb-a66a-646eec2192df.mov

The result is that two tracks are now inserted at the same time. But, the timing of the popup warning is still wrong.

After this PR makes elimination of judging by drawing frame and setting the queue manually, all tracks will be inserted correctly at the same time with the same code except for setting the queue manually.

https://user-images.githubusercontent.com/61938263/130363536-5870330e-7a37-4b00-9740-05b145acbb2a.mov

File: editor/animation_track_editor.cpp [hunk 4]
```diff
@@ -3383,74 +3369,82 @@ void AnimationTrackEditor::_query_insert(const InsertData &p_id) {
...
+	for (const InsertData &E : insert_data) {
+		// Prevent insertion of multiple tracks
+		if (E.path == p_id.path) {
+			return; // Already inserted a track for this on this frame
+		}
+	}
+
+	insert_data.push_back(p_id);
+
+	// Without queue, commit immidiately.
```
Comment (fire) [inline editor/animation_track_editor.cpp:3441]: Grammar.

> // Without queue, commit immidiately.

> // Without queue, commit immediately.

File: editor/animation_track_editor.cpp [hunk 5]
```diff
@@ -3383,74 +3369,82 @@ void AnimationTrackEditor::_query_insert(const InsertData &p_id) {
...
-		return;
+void AnimationTrackEditor::_query_insert(const InsertData &p_id) {
+	if (!insert_queue) {
+		insert_data.clear();
 	}
 
+	for (const InsertData &E : insert_data) {
+		// Prevent insertion of multiple tracks
+		if (E.path == p_id.path) {
+			return; // Already inserted a track for this on this frame
```
Comment (fire) [inline editor/animation_track_editor.cpp:3435]: > // Already inserted a track for this on this frame

> // Already inserted a track this frame.

File: editor/animation_track_editor.cpp [hunk 6]
```diff
@@ -3383,74 +3369,82 @@ void AnimationTrackEditor::_query_insert(const InsertData &p_id) {
...
+	}
 
-			insert_confirm_bezier->set_visible(all_bezier);
-			insert_confirm_reset->set_visible(reset_allowed);
-
-			insert_confirm->get_ok_button()->set_text(TTR("Create"));
-			insert_confirm->popup_centered();
-			insert_query = true;
+	if (bool(EDITOR_DEF("editors/animation/confirm_insert_track", true)) && num_tracks > 0) {
+		//potential new key, does not exist
```
Comment (fire) [inline editor/animation_track_editor.cpp:3407]: > //potential new key, does not exist

> // Potentially a new key, does not exist.

Can you check the other comments?

**Comment by TokageItLab** [comment]
#45699 has been merged, so updated it. Currently the animation track insertion by SkeletonEditor is broken, but it will be fixed when this PR is merged.

**Review by akien-mga — APPROVED** [review]

**Pull request merged**